search contexts: non-functional dropdown menu

limitedmage opened the pull request:
Dropdown menu just for styling. Data and functionality will be added later.
Text in the menu is cut off with ellipses but will show when hovered.

Note @quinnkeast: hover style doesn't match Figma but does match our existing hover style for dropdown menus.

![Screen Shot 2021-02-05 at 2 02 43 PM](https://user-images.githubusercontent.com/206864/107094036-2effe680-67bb-11eb-954e-c21383fc9e22.png)

quinnkeast reviewed (commented):
I've made a few suggestions with the changes that'll fine-tune the design implementation. Hopefully they're in line with our style and conventions, but if not, at least the intent is clear!

Some other notes:

- I assume this is because there's no functionality yet, but the currently-selected context should have a darker background to distinguish between "currently active context" and "currently selected item in list." Assuming this is what you referenced about the existing hover style for dropdown menus, this is important for getting the interaction behaviour and visual affordances correct.
- The focus state for the selected context shouldn't have an outline, just the background color change. This is equivalent to the autocomplete suggestions.
- The "Default" badge is perfect for when it has a highlighted background (currently selected or hovered), and should also have `--color-bg-2` when the row is not highlighted. (This is missing from the design master—I'm sorry! I didn't move it over from the working section.)
- The "Find a context" input should autofocus whenever the dropdown is opened. 
- While the "Find a context" input is focused, the down arrow key should always move selection to the list items. Likewise, if the user has selected the top item in the list of contexts, the update arrow should re-focus the "Find a context" input. (May be part of functionality—if so, ignore!)
- Each description should have padding on the right such that the "Default" badge has its own "column," and never overlaps the content.
- Should tab focusing also focus on Reset / Manage contexts? I think yes, because of accessibility needs combined with the interaction behaviour meaning that the user shouldn't be tabbing from the dropdown to the search query input—they'll be selecting a context or hitting "esc," both of which close the dropdown and jump to the search query input.

Since creating the original design, we decided on the slightly updated approach to namespacing. The original design (`contextName | description`) works well for the automatic contexts we're implementing for Cloud, but has some challenges for fully namespaced contexts (`@longusername/co... | description`). I'm going to do a new design exploration to figure out how to better handle the context names in this dropdown now that we have the namespacing. I don't think this should be a blocker for the current work, because it won't be relevant until we start working on Enterprise contexts (adapting version contexts) / custom contexts.

limitedmage commented:
Thanks @quinnkeast! I've addressed most of your feedback but have some comments about the following:


> * I assume this is because there's no functionality yet, but the currently-selected context should have a darker background to distinguish between "currently active context" and "currently selected item in list." Assuming this is what you referenced about the existing hover style for dropdown menus, this is important for getting the interaction behaviour and visual affordances correct.

Ahh that makes more sense! I'll make sure to take that into account when styling the selected state.

> * The focus state for the selected context shouldn't have an outline, just the background color change. This is equivalent to the autocomplete suggestions.

This does not match our current styling for other dropdowns (eg. the user menu). Dropdown menu items have an outline and I think we should keep it for accessibility purposes.

> * The "Find a context" input should autofocus whenever the dropdown is opened.
> * While the "Find a context" input is focused, the down arrow key should always move selection to the list items. Likewise, if the user has selected the top item in the list of contexts, the update arrow should re-focus the "Find a context" input. (May be part of functionality—if so, ignore!)
> * Should tab focusing also focus on Reset / Manage contexts? I think yes, because of accessibility needs combined with the interaction behaviour meaning that the user shouldn't be tabbing from the dropdown to the search query input—they'll be selecting a context or hitting "esc," both of which close the dropdown and jump to the search query input.

This is not built into the control so it will take a bit of effort to build. In particular, Bootstrap dropdowns have great keyboard accessibility for the menu items themselves but are not really meant to hold other kinds of controls like links or input boxes. I've noted a separate work item for this: https://github.com/orgs/sourcegraph/projects/113#card-54511979

> * Each description should have padding on the right such that the "Default" badge has its own "column," and never overlaps the content.

This would severely cut into the little space there is in the dropdown, as only one line item will have the "Default" badge. I don't know if this is a good idea. I think the badges should look like the labels in places like Gmail where items without them still render in the whole space without leaving any whitespace.

> Since creating the original design, we decided on the slightly updated approach to namespacing. The original design (`contextName | description`) works well for the automatic contexts we're implementing for Cloud, but has some challenges for fully namespaced contexts (`@longusername/co... | description`). I'm going to do a new design exploration to figure out how to better handle the context names in this dropdown now that we have the namespacing. I don't think this should be a blocker for the current work, because it won't be relevant until we start working on Enterprise contexts (adapting version contexts) / custom contexts.

quinnkeast commented:
> > * The focus state for the selected context shouldn't have an outline, just the background color change. This is equivalent to the autocomplete suggestions.
> 
> This does not match our current styling for other dropdowns (eg. the user menu). Dropdown menu items have an outline and I think we should keep it for accessibility purposes.

This is the one part I need to push back on. I see your case, but I don't think it's right to think of this element as a strict "dropdown" in the same sense the menus are dropdowns—instead, it's a self-contained UI element within a larger dropdown. Using the prominent outline for the selected context makes the overall UI less clear by changing the visual prominence in a way that's not intended. We still have visual feedback around the selected state with the background change, so it's not made inaccessible by removing the glow—it's just an alternate way of highlighting the current selection.

A close UI parallel on GitHub would be the branch selector on a repository, which can be toggled, searched, and navigated with just the keyboard:

![CleanShot 2021-02-09 at 11 28 12](https://user-images.githubusercontent.com/6304497/107350625-f5382400-6ac9-11eb-8447-0ed7f18805e2.gif)

> > * The "Find a context" input should autofocus whenever the dropdown is opened.
> > * While the "Find a context" input is focused, the down arrow key should always move selection to the list items. Likewise, if the user has selected the top item in the list of contexts, the update arrow should re-focus the "Find a context" input. (May be part of functionality—if so, ignore!)
> > * Should tab focusing also focus on Reset / Manage contexts? I think yes, because of accessibility needs combined with the interaction behaviour meaning that the user shouldn't be tabbing from the dropdown to the search query input—they'll be selecting a context or hitting "esc," both of which close the dropdown and jump to the search query input.
> 
> This is not built into the control so it will take a bit of effort to build. In particular, Bootstrap dropdowns have great keyboard accessibility for the menu items themselves but are not really meant to hold other kinds of controls like links or input boxes. In the effort of keeping this PR small, I've split this a separate work item: https://github.com/orgs/sourcegraph/projects/113#card-54511979
> 

Sounds good – thank you! This is a pretty crucial part of the interaction, so I'm happy to spent the effort on it. Similar to above, I think it's important we don't think of this as conceptually equivalent to a dropdown menu.

> > * Each description should have padding on the right such that the "Default" badge has its own "column," and never overlaps the content.
> 
> This would severely cut into the little space there is in the dropdown, as only one line item will have the "Default" badge. I don't know if this is a good idea. I think the badges should look like the labels in places like Gmail where items without them still render in the whole space without leaving any whitespace.

Sounds good – let's do it. The only thing I'd want to be sure of is that the badge has a graceful "seam" with the description, and doesn't cut off text in a way that looks broken. I see that Gmail's labels on hover always clip the text with `...` to suit whichever content is being show on hover:

<img width="288" alt="CleanShot 2021-02-09 at 11 23 50@2x" src="https://user-images.githubusercontent.com/6304497/107350155-5ca1a400-6ac9-11eb-8539-3dd4a21bb31c.png">

If we can accomplish similar here, that'd be stellar.

felixfbecker commented:
Just wanna say I love seeing the attention to detail and accessibility here. And thank you @limitedmage for making sure we don't implement this as a one-off, I'm sure we'll need these kinds of components more elsewhere. Might want to involve @sourcegraph/frontend-platform on this an kick off a separate discussion on that.

Something I got reminded of reading the above that I wanted to share that Bootstrap dropdowns are explicitly _not_ "menus" as defined by ARIA `role="menu"`, because those can only contain menu items, checkbox menu items, radio button menu items, radio button groups, and sub-menus. Bootstrap's dropdowns are explicitly designed to support other arbitrary interactive content too (like the rev switcher shown above from GitHub with an input). This is in opposite to `@reach/menu`'s dropdown, which sticks strictly to the ARIA definition.

> The WAI ARIA standard defines an actual role="menu" widget, but this is specific to application-like menus which trigger actions or functions. ARIA menus can only contain menu items, checkbox menu items, radio button menu items, radio button groups, and sub-menus.
> 
> Bootstrap’s dropdowns, on the other hand, are designed to be generic and applicable to a variety of situations and markup structures. For instance, it is possible to create dropdowns that contain additional inputs and form controls, such as search fields or login forms. For this reason, Bootstrap does not expect (nor automatically add) any of the role and aria- attributes required for true ARIA menus. Authors will have to include these more specific attributes themselves.
> 
> However, Bootstrap does add built-in support for most standard keyboard menu interactions, such as the ability to move through individual .dropdown-item elements using the cursor keys and close the menu with the ESC key.

See https://getbootstrap.com/docs/5.0/components/dropdowns/#accessibility

5h1rU commented on client/web/src/search/input/SearchContextDropdown.scss:
```diff
@@ -27,6 +27,11 @@
+
+        &::after {
+            // Hide chevron usually shown by dropdown buttons
+            display: none;
```
wondering if this is the equivalent to include `caret={false}` as prop on `DropdownToggle` ?

limitedmage replied:
Unfortunately this doesn't work in combination with the `dropdown-toggle` class. Without the class the border radius is messed up and I have to override that instead. This is easier to override.

limitedmage commented:
@quinnkeast:
Re: dropdown menu item focus style:

> A close UI parallel on GitHub would be the branch selector on a repository

I don't think GitHub's dropdown menus are a good parallel because their menus don't have an outline for the focused items while ours do. GitHub's repo selector is the same control as other dropdown menus on their site (eg. the "Watch" dropdown). In fact, GitHub seems to have **multiple** different dropdown styles for different menus with no apparent consistency :S We should strive for our webapp have consistent styles.

Re: cutting off text with default label

> The only thing I'd want to be sure of is that the badge has a graceful "seam" with the description, and doesn't cut off text in a way that looks broken.

Thanks for pointing this out, I added a small left margin to avoid this:

![image](https://user-images.githubusercontent.com/206864/107398815-de3b0700-6ab4-11eb-9fcf-904beb89f66c.png)
